lib/Middleware.php

```diff
@@ -2,6 +2,14 @@
 
 namespace Aerys;
 
+use Amp\Promise;
+
 interface Middleware {
-    public function do(InternalRequest $ireq);
+    /**
+     * @param \Aerys\Request $request
+     * @param Responder $responder Request responder.
+     *
+     * @return \Amp\Promise<\Aerys\Response>
+     */
+    public function process(Request $request, Responder $responder): Promise;
```

Comment:
wait … are we now really removing the access to internal structures here? :x (Request instead of Internal\Request)
I found that a rather powerful feature, when necessary. Not that _most_ use cases would need that, but it may be useful in some cases...

Also, that way you essentially removed the possibility to mutate the request before the responder can see it. Sure, I could fake a request with a my own Internal\Request instance, but, no. No.

Reply:
No, you can still mutate the request before delegating to the responder, just like middlewares work in PSR-15. Which use cases do you have in mind?

Reply:
Well, then please show me how (while still fulfilling the interface contracts).

Reply:
Request is supposed to be mutable, but isn't yet. I guess that's on @trowski's TODO list. We discussed that earlier.

Reply:
@bwoebi I planned on adding options to mutate `Aerys\Request`, exposing only the things that should be mutated.